File: answer_match.cpp
#include<cstdio>
const int buf_sz=439999,answer_cnt=3999,str_len=999;
char buf[buf_sz];
int p[answer_cnt][7],f[str_len][str_len];
inline int max(int a,int b){return a>b?a:b;}
inline int lcs(int*_a,int*_b){
	char*a=buf+*_a,*b=buf+*_b;
	int al=_a[1]-*_a,bl=_b[1]-*_b;
	if(!al||!bl)return 0;
	for(int i=0;i<=al;i++)f[i][0]=0;
	for(int i=0;i<=bl;i++)f[0][i]=0;
	for(int i=0;i<al;i++)for(int j=0;j<bl;j++)
		f[i+1][j+1]=a[i]==b[j]?f[i][j]+1:max(f[i][j+1],f[i+1][j]);
	return f[al][bl];
}
int main(){
	//FILE*ans=fopen("answers.txt","r");
	int sz,n=0,score=-1,best,t,v;
	sz=fread(buf,1,sizeof buf,stdin);
	//buf[sz++]='\n';
	//sz+=fread(buf+sz,1,sizeof buf-sz,ans);
	for(int i=0;i<sz;i++)if(buf[i]=='\n')p[++n][0]=i;
	p[++n][0]=sz;
	for(int i=0;i<n;i++){
		for(int j=1;j<7;j++)p[i][j]=p[i+1][0];
		for(int j=p[i][0],c=0;c<6&&j<p[i+1][0];j++)
			if(buf[j]==',')p[i][++c]=j;
	}
	for(int i=1;i<n;i++){
		t=lcs(p[0],p[i]);
		for(int j=0;v=0,j<5;j++,t+=v)for(int k=0;k<5;k++)
			v=max(v,lcs(p[0]+j,p[i]+k));
		if(t>score)score=t,best=i;
	}
	if(score==0)return puts("0"),0;
	v=best,score=-1;
	for(int i=1;i<5;i++){
		t=lcs(p[0]+i,p[v]+5);
		if(t>score)score=t,best=i;
	}
	printf("%d\n",best);
}

File: answer_match.py
import os,threading
import log
import environment
from subprocess import Popen,PIPE

lock=threading.RLock()
plock=threading.RLock()

def calLCS(a,b):
	a=list(a);b=list(b)
	n=len(a);m=len(b)
	dp=[[0 for j in xrange(m+1)]for i in xrange(n+1)]
	for i in xrange(1,n+1):
		for j in xrange(1,m+1):
			dp[i][j]=max(dp[i-1][j],dp[i][j-1])
			if i and j and(a[i-1]==b[j-1]):
				dp[i][j]=dp[i-1][j-1]+1
	return dp[n][m]

def getBestAns1(x):
	global q
	plock.acquire()
	log.log('get best ans')
	plock.release()
	lock.acquire()
	score=-1
	x=map(unicode,x)
	for y in q:
		t=calLCS(x[0],y[0])
		for i in range(1,5):
			v=0
			for j in range(1,5):
				v=max(v,calLCS(x[i],y[j]))
			t+=v
		if t>score:
			score=t
			best=y
	lock.release()
	if score==-1:return 0
	score=-1;ans=0
	for i in range(1,5):
		t=calLCS(x[i],best[5])
		if t>score:
			score=t
			ans=i
	if score<=0:return 0
	return ans

def getBestAns2(x):
	global q,qs
	plock.acquire()
	log.log('get best ans')
	plock.release()
	lock.acquire()
	t=x[0]+','+x[1]+','+x[2]+','+x[3]+','+x[4]
	cmd='answer_match'
	if not environment.is_windows:
		cmd='./'+cmd
	p=Popen(cmd,stdin=PIPE,stdout=PIPE)
	p.stdin.write((t+'\n'+qs).encode('utf-8','ignore'))
	p.stdin.close()
	res=int(p.stdout.readline())
	lock.release()
	return res

def getBestAns(x):
	return getBestAns2(x)

q=[]
with open('answers.txt','r') as f:
	t=f.readlines()
	qs=''
	for i in t:
		qs+=i.decode('utf-8')
		#q.append(i.decode('utf-8'))
		v=i.decode('utf-8').strip('\n').strip('\r').split(',')
		if len(v)==6:q.append(v)

'''
x=[]
for i in range(5):
	x.append(raw_input().decode('gbk'))
print getBestAns(x)
'''
